### PR TITLE
Implement instance users management

### PR DESCRIPTION
**Pull request by allenwq**

### CODE REVIEW

File: app/controllers/system/admin/instance/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+class System::Admin::Instance::UsersController < System::Admin::Instance::Controller
+  load_and_authorize_resource :instance_user, class: InstanceUser.name,
```
Comment (lowjoel) [inline app/controllers/system/admin/instance/users_controller.rb]: By convention, this should be :user

File: app/controllers/system/admin/instance/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+                                              parent: false, except: [:index]
+  add_breadcrumb :index, :admin_instance_users_path
+
+  def index
+    @instance_users = InstanceUser.includes(user: [:emails, :courses]).
```
Comment (lowjoel) [inline app/controllers/system/admin/instance/users_controller.rb]: You should probably use @instance.users.includes...

Comment (lowjoel) [inline app/controllers/system/admin/instance/users_controller.rb]: Assume that acts_as_tenant is disabled.

File: app/models/components/ability_host.rb [hunk 3]
```diff
@@ -44,8 +44,37 @@ def course_managers_hash
...
+        instance_users[:role] = roles.map do |role|
+          [InstanceUser.roles[role], role.to_s]
+        end.flatten!
+      end
+
+      { instance_users: instance_users }
+    end
+
+    # @return [Hash] The hash is relative to a component which has a +belongs_to+ association with
+    #   a Instance.
```
Comment (lowjoel) [inline app/models/components/ability_host.rb]: an Instance

File: app/models/components/ability_host.rb [hunk 4]
```diff
@@ -44,8 +44,37 @@ def course_managers_hash
     end
   end
 
+  module InstanceHelpers
+    protected
+
+    # Creates a hash which allows referencing a set of instance users.
+    #
+    # @param [Array<Symbol>] roles The roles {InstanceUser::Roles} which should be referenced by
+    # this rule.
```
Comment (lowjoel) [inline app/models/components/ability_host.rb]: spaces before this.

File: app/models/components/ability_host.rb [hunk 5]
```diff
@@ -44,8 +44,37 @@ def course_managers_hash
...
+    # @return [Hash] This hash is relative to a Instance.
+    def instance_user_hash(*roles)
+      instance_users = { user_id: user.id }
+      unless roles.empty?
+        instance_users[:role] = roles.map do |role|
+          [InstanceUser.roles[role], role.to_s]
```
Comment (lowjoel) [inline app/models/components/ability_host.rb]: I hope we can go to Rails 5 soon :(

File: app/models/instance_user.rb [hunk 6]
```diff
@@ -3,4 +3,6 @@ class InstanceUser < ActiveRecord::Base
 
   enum role: { normal: 0, instructor: 1, administrator: 2 }
   belongs_to :user, inverse_of: :instance_users
+
+  scope :ordered_by_username, -> { order('users.name asc') }
```
Comment (lowjoel) [inline app/models/instance_user.rb]: I rather not use strings. Use hashes so that renaming of tables by aliases can be dealt with by Rails.

**Comment by lowjoel** [comment]
Please rebase after fixing my comments.

File: app/controllers/system/admin/instance/users_controller.rb (hunk 1 above)
Comment (allenwq) [inline app/controllers/system/admin/instance/users_controller.rb]: Eh, don't you think that `@user.user` is confusing ?  Here I'm actually fetching the instance_users but present `instance_user.user`

Comment (lowjoel) [inline app/controllers/system/admin/instance/users_controller.rb]: then it should be InstanceUsersController.

Comment (allenwq) [inline app/controllers/system/admin/instance/users_controller.rb]: Then it's not properly namespaced.

I think this controller is similar to `Course::UsersController`?  In `Course::UsersController` we are also fetching `@course_user` and present `@course_user.user`

Comment (lowjoel) [inline app/controllers/system/admin/instance/users_controller.rb]: System::Admin::Instance::InstanceUsersController

Comment (lowjoel) [inline app/controllers/system/admin/instance/users_controller.rb]: If the other one needs to be fixed then maybe we should fix that.

File: app/models/instance_user.rb (hunk 6 above)
Comment (allenwq) [inline app/models/instance_user.rb]: Seem`order` does not support hash expressions when referring another table (in this case `users`) ?

**Pull request merged**